### PR TITLE
Feat: sharing account

### LINKED ISSUE #651: Ability to create Share Account for clients
**Summary:** 

Summarize your issue in one sentence (what goes wrong, what did you expect to happen)

**Steps to reproduce:** 

How can we reproduce the issue?

**Expected behavior:** 

What did you expect the app to do?

**Observed behavior:** 

What did you see instead?  Describe your issue in detail here.

**Device and Android version:** 

What make and model device (e.g., Samsung Galaxy S3) did you encounter this on?  What Android
version (e.g., Android 4.0 Ice Cream Sandwich or Android 6.0 Marshmallow) are you running?  Is it
 the stock
version from the manufacturer or a custom ROM?

**Screenshots:** 

Can be created by pressing the Volume Down and Power Button at the same time on Android 4.0 and higher.

### PR DESCRIPTION
**Pull request by mayank-kgp**
Resolves #651 
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Apply the `MifosStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            holder.tvChargeName.setText(chargeOptionList.get(position).getName());
+            holder.etChargeAmount.setText(String.valueOf(chargeOptionList.
+                    get(position).getAmount()));
+
+            holder.btnRemove.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    chargeOptionList.remove(position);
+                    notifyItemRemoved(position);
+                    notifyDataSetChanged();
```
Comment (droidchef) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java]: Why calling ```notifyDataSetChanged()``` ?

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    private Context c;
+    private List<ChargeOptions> chargeOptionList = new ArrayList<ChargeOptions>();
+
+    @Inject
+    public ShareChargesAdapter(@ActivityContext Context context) {
+        c = context;
+    }
+
+    public List<ChargeOptions> getChargeOptionList() {
```
Comment (droidchef) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java]: Why are we exposing this list with this public getter ?

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        return chargeOptionList;
+    }
+
+    public void addChargeOptionList(ChargeOptions chargeOption) {
+        chargeOptionList.add(chargeOption);
```
Comment (droidchef) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java]: What is this method doing here ?

File: mifosng-android/src/main/java/com/mifos/objects/organisation/ProductSharing.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * Created by mayankjindal on 18/08/17.
+ */
+
+public class ProductSharing implements Parcelable {
```
Comment (droidchef) [inline mifosng-android/src/main/java/com/mifos/objects/organisation/ProductSharing.java]: Annotate all your fields with ```@Nullable``` or ```@NonNull``` annotations properly.

File: mifosng-android/src/main/java/com/mifos/api/datamanager/DataManagerSharing.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * Created by mayankjindal on 18/08/17.
+ */
+/**
+ * This DataManager is for Managing Client's Share Account, In which Request is going to Server
+ * and In Response, We are getting Client's Share Account Observable Response using Retrofit2 .
+ */
+public class DataManagerSharing {
+
+    public final BaseApiManager mBaseApiManager;
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/api/datamanager/DataManagerSharing.java]: I know at many places we are using `m` convention in naming but we no longer supporting this convention please rename the variable `baseApiManager`

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+        args.putInt(Constants.CLIENT_ID, clientId);
+        sharingAccountFragment.setArguments(args);
+        return sharingAccountFragment;
+    }
+
+    @Override
+    public void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        ((MifosBaseActivity) getActivity()).getActivityComponent().inject(this);
+        if (getArguments() != null)
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: Use braces to enclose the `if` statement

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+                .simple_spinner_item, mListSavingsAccountOptions);
+        mSavingsAccountAdapter.setDropDownViewResource(
+                android.R.layout.simple_spinner_dropdown_item);
+        spSavingsAccount.setAdapter(mSavingsAccountAdapter);
+        spSavingsAccount.setOnItemSelectedListener(this);
+    }
+
+    @OnClick(R.id.btn_submit)
+    void submitSharingAccount() {
+        if (Network.isOnline(getContext())) {
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: I don't think we should use `getContext()` we should use `getActivity()`. what do you say @droidchef

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+    @OnClick(R.id.tv_submittedon_date)
+    void onClickTextViewSubmissionDate() {
+        isSubmissionDate = true;
+        mfDatePicker.show(getActivity().getSupportFragmentManager(), FragmentConstants
+                .DFRAG_DATE_PICKER);
+    }
+
+    public void inflateApplicationDate() {
+        mfDatePicker = MFDatePicker.newInsance(this);
+
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: Remove this extra line

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+        mListSavingsAccountOptions.addAll(mSharingAccountPresenter
+                .filterSavingProductsNames(sharingAccountsTemplate.getClientSavingsAccounts()));
+        mSavingsAccountAdapter.notifyDataSetChanged();
+        showDefaultValues();
+    }
+
+    @Override
+    public void showSharingAccountCreatedSuccessfully(ShareResponse shareResponse) {
+        Toast.makeText(getActivity(),
+                getResources().getString(R.string.sharing_account_submitted_for_approval),
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: don't need to call `getResources()`, just call R.string.sharing_account_submitted_for_approval.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+                Toast.LENGTH_LONG).show();
+        getActivity().getSupportFragmentManager().popBackStackImmediate();
+    }
+
+    @Override
+    public void showFetchingError(int errorMessage) {
+        Toaster.show(rootView, getResources().getString(errorMessage));
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: use `getString(errorMessage)`

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountPresenter.java [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+    @Override
+    public void attachView(SharingAccountMvpView mvpView) {
+        super.attachView(mvpView);
+    }
+
+    @Override
+    public void detachView() {
+        super.detachView();
+        mSubscriptions.unsubscribe();
```
Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountPresenter.java]: @mayank-kgp use `mSubscriptions.clear()`

File: mifosng-android/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -160,6 +164,13 @@
     <string name="group_clients">Group Clients</string>
     <string name="identifier_deleted_successfully">Identifier deleted Successfully</string>
     <string name="exact_match">Exact Match</string>
+    <string name="share_charges">Charges</string>
+    <string name="requested_shares">Shares</string>
+    <string name="lockin_period_frequency">Lock In Period Frequency</string>
+    <string name="lockin_period">Lock In Period</string>
+    <string name="min_active_period_frequency">Minimum Active Period Frequency</string>
+    <string name="min_active_period">Minimum Active Period</string>
+    <string name="saving_account">Savings Account</string>
```
Comment (therajanmaurya) [inline mifosng-android/src/main/res/values/strings.xml]: @mayank-kgp please look this string https://github.com/mayank-kgp/android-client/blob/e3abac9455e661501d555121e08eca5baa7a14eb/mifosng-android/src/main/res/values/strings.xml#L316 It's the same value

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java (hunk 9 above)
Comment (droidchef) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: @therajanmaurya does ```getString``` work for lower API versions also ?

Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: yes It was added in API 1 : https://developer.android.com/reference/android/content/Context.html#getString(int)

Comment (therajanmaurya) [inline mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: Please look the answer here too https://stackoverflow.com/a/20539170/3176277

**Comment by mayank-kgp** [comment]
@therajanmaurya Please see the UI.

shareaccount

File: mifosng-android/src/main/java/com/mifos/api/model/APIEndPoint.java [hunk 13]
```diff
@@ -21,6 +21,7 @@
     public static final String SEARCH = "search";
     public static final String CREATESAVINGSACCOUNTS = "savingsaccounts";
     public static final String CREATESAVINGSPRODUCTS = "savingsproducts";
+    public static final String CREATESHARINGACCOUNTS = "accounts/share";
```
[the remark this answers is not in the document]
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/api/model/APIEndPoint.java]: Done.

File: mifosng-android/src/main/java/com/mifos/api/services/SharingService.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+     * This Service is for fetching the SharingAccountsTemplate.
+     * REST END POINT:
+     * https://demo.openmf.org/fineract-provider/api/v1/accounts/share/template?clientId={clientId}&productId={productId}
+     *
+     * @param clientId Client Id
+     * @return SharingProductTemplate
+     */
+    @GET(APIEndPoint.CREATESHARINGACCOUNTS + "/template")
+    Observable<SharingAccountsTemplate> getClientSharingAccountTemplateByProduct(
+            @Query("clientId") int clientId, @Query("productId") int productId);
```
[the remark this answers is not in the document]
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/api/services/SharingService.java]: We had. I have replaced strings.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+import butterknife.ButterKnife;
+
+/**
+ * Created by mayankjindal on 21/08/17.
+ */
+
+public class ShareChargesAdapter extends
+        RecyclerView.Adapter<ShareChargesAdapter.ViewHolder> {
+
+    private Context c;
```
[the remark this answers is not in the document]
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java]: Done.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java (hunk 2 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java]: Removed.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java (hunk 3 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java]: changed.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java (hunk 1 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/adapters/ShareChargesAdapter.java]: changed.

File: mifosng-android/src/main/java/com/mifos/objects/organisation/ProductSharing.java (hunk 4 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/objects/organisation/ProductSharing.java]: Annotated.

File: mifosng-android/src/main/java/com/mifos/api/datamanager/DataManagerSharing.java (hunk 5 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/api/datamanager/DataManagerSharing.java]: Done.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java (hunk 6 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: done.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java (hunk 7 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: I have changed it to getAcitivity().

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java (hunk 8 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: done.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java (hunk 9 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: Removed getResources().

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java (hunk 10 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountFragment.java]: done.

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountPresenter.java (hunk 11 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/java/com/mifos/mifosxdroid/online/sharingaccount/SharingAccountPresenter.java]: done.

File: mifosng-android/src/main/res/values/strings.xml (hunk 12 above)
Comment (mayank-kgp) [reply mifosng-android/src/main/res/values/strings.xml]: done.

**Comment by tarun0** [comment]
Closing for the lack of updates.

**Pull request closed without merging**